src/main/java/org/cloudfoundry/promregator/lite/discovery/DiscoveryService.kt

```diff
@@ -0,0 +1,48 @@
+package org.cloudfoundry.promregator.lite.discovery
+
+import com.github.benmanes.caffeine.cache.Caffeine
+import kotlinx.coroutines.future.await
+import kotlinx.coroutines.runBlocking
+import org.cloudfoundry.promregator.lite.config.PromregatorConfigurationV2
+import org.cloudfoundry.promregator.lite.discovery.SimpleAppInstanceScanner.Instance
+import org.springframework.stereotype.Service
+import java.util.concurrent.TimeUnit
+
+private val log = mu.KotlinLogging.logger {}
+
+@Service
+class DiscoveryService(
+        private val promregatorConfiguration: PromregatorConfigurationV2,
+        private val targetResolver: SimpleTargetResolver,
+        private val appInstanceScanner: SimpleAppInstanceScanner,
+) {
+    val cache = Caffeine.newBuilder()
+        .maximumSize(10_000)
+        .expireAfterWrite(60, TimeUnit.MINUTES)
+        .refreshAfterWrite(60, TimeUnit.SECONDS)
+        .buildAsync {_: String -> runBlocking { discover() }}
+
+    suspend fun discoverCached(): List<Instance>  {
+        return cache.get("default").await()
+    }
+
+    suspend fun discover(): List<Instance> {
```

Comment:
The new `discover` is also simpler since all CF API calls happen in the `resolveTargets` call. The `determineInstancesFromTargets` can eventually be moved into this class, however I kept it in the old class for now to simplify the comparison discussion.